implement AccIsEnabled

SimeonEhrig opened the pull request:
The value of AccIsEnabled indicates whether an accelerator is enabled for a specific tag.

fwyzard commented on include/alpaka/acc/Tag.hpp:
```diff
@@ -56,3 +57,32 @@ namespace alpaka
     inline constexpr bool accMatchesTags = (std::is_same_v<alpaka::AccToTag<TAcc>, TTag> || ...);
 } // namespace alpaka
+
+// include all Acc's because of the struct AccIsEnabled
+// if an acc is not include, it will be not enabled independent of the compiler flags
+// include after defining the traits alpaka::AccToTag and alpaka::TagToAcc because it
+// is a circular dependency
+#include "alpaka/acc/AccCpuOmp2Blocks.hpp"
+#include "alpaka/acc/AccCpuOmp2Threads.hpp"
+#include "alpaka/acc/AccCpuSerial.hpp"
+#include "alpaka/acc/AccCpuSycl.hpp"
+#include "alpaka/acc/AccCpuTbbBlocks.hpp"
+#include "alpaka/acc/AccCpuThreads.hpp"
+#include "alpaka/acc/AccFpgaSyclIntel.hpp"
+#include "alpaka/acc/AccGpuCudaRt.hpp"
+#include "alpaka/acc/AccGpuHipRt.hpp"
+
+namespace alpaka
+{
+    /// @brief check if the accelerator is enabled for a given tag
+    /// @tparam TTag alpaka tag type
+    template<typename TTag, typename = void>
+    struct AccIsEnabled : std::false_type
+    {
+    };
+
+    template<typename TTag>
+    struct AccIsEnabled<TTag, std::void_t<TagToAcc<TTag, alpaka::DimInt<1>, int>>> : std::true_type
+    {
+    };
+} // namespace alpaka
```
Would it be safer to move this to a separate header file, _e.g._ `AccIsEnabled.h`, that can include all the others with out risk of a circular dependency ?

SimeonEhrig replied:
Your are right. I fixed it.

mehmetyusufoglu commented:
> The value of AccIsEnabled indicates whether an accelerator is enabled for a specific tag.

[question] Where will this be used?

SimeonEhrig commented:
> > The value of AccIsEnabled indicates whether an accelerator is enabled for a specific tag.
> 
> [question] Where will this be used?

To make it easier to write tests. An example from PR #2180 

```c++
using TagList = std::tuple<
    alpaka::TagCpuSerial,
    alpaka::TagCpuThreads,
    alpaka::TagCpuTbbBlocks,
    alpaka::TagCpuOmp2Blocks,
    alpaka::TagCpuOmp2Threads,
    alpaka::TagGpuCudaRt,
    alpaka::TagGpuHipRt,
    alpaka::TagCpuSycl,
    alpaka::TagFpgaSyclIntel,
    alpaka::TagGpuSyclIntel>;

using TagTagList = alpaka::meta::CartesianProduct<std::tuple, TagList, TagList>;

TEMPLATE_LIST_TEST_CASE("printDefines", "[mem][visibility]", TagTagList)
{
    using Tag1 = std::tuple_element_t<0, TestType>;
    using Tag2 = std::tuple_element_t<1, TestType>;

    if constexpr(alpaka::AccIsEnabled<Tag1>::value && alpaka::AccIsEnabled<Tag2>::value)
    {
        SUCCEED(Tag1::get_name() << " + " << Tag2::get_name());

        using Acc1 = alpaka::TagToAcc<Tag1, Dim, Idx>;
        using Acc2 = alpaka::TagToAcc<Tag2, Dim, Idx>;
        // ...

        if constexpr((isCPUTag<Tag1>() && isCPUTag<Tag2>()) || std::is_same_v<Tag1, Tag2>)
        {
            STATIC_REQUIRE(alpaka::hasSameMemView(dev1, bufDev2));
            STATIC_REQUIRE(alpaka::hasSameMemView(dev2, bufDev1));
        }
        else
        {
            STATIC_REQUIRE_FALSE(alpaka::hasSameMemView(dev1, bufDev2));
            STATIC_REQUIRE_FALSE(alpaka::hasSameMemView(dev2, bufDev1));
        }
}
```

SimeonEhrig commented:
During the VC today, @psychocoderHPC gave some improvements for this PR:
- [x] add the `TagList` to the official API and rename it to `AccTags`
- [x] add function, which returns list of tags, where the acc is enabled
- [x] add trait, which return list of tags, which are bonded to Acc's which uses the `cpuplatform`
- change all tests to use tags instead the `EnabledAccsElseInt` list (extra PR)

psychocoderHPC commented on include/alpaka/acc/Tag.hpp:
```diff
@@ -55,4 +55,36 @@ namespace alpaka
     template<typename TAcc, typename... TTag>
     inline constexpr bool accMatchesTags = (std::is_same_v<alpaka::AccToTag<TAcc>, TTag> || ...);
+
+    // list of all available tags
```
please use `//!` to create a doxygen comment

psychocoderHPC commented on include/alpaka/acc/TagAccIsEnabled.hpp:
```diff
@@ -0,0 +1,36 @@
+    {
+    };
+
+    // list of all tags where the related accelerator is enabled
```
please use `//!` to create a doxygen comment

psychocoderHPC commented on include/alpaka/acc/Tag.hpp:
```diff
@@ -55,4 +55,36 @@ namespace alpaka
+        alpaka::TagFpgaSyclIntel,
+        alpaka::TagGpuSyclIntel>;
+
+    /// @brief check if the accelerator related to the tag is bounded to the cpu platform
```
please sue `\` for doxygen comment keywords

psychocoderHPC commented:
short doxygen introduction: https://www.doxygen.nl/manual/docblocks.html

SimeonEhrig commented:
@psychocoderHPC ready to merge

psychocoderHPC commented on include/alpaka/acc/Tag.hpp:
```diff
@@ -41,18 +41,52 @@ namespace alpaka
+        alpaka::TagFpgaSyclIntel,
+        alpaka::TagGpuSyclIntel>;
+
+    //! \brief check if the accelerator related to the tag is bounded to the cpu platform
```
we should reformulate `bounded to the cpu platform` because `TagCpuSycl` is not using the alpaka CPU platform, it has its own platform.

I think you have already your memory visibilities in mind and you like to express that you have access to the same kind of memory.
Let us think tomorrow in the meeting together how we can call this and what it should say.

psychocoderHPC replied:
discussed offline: this trait will be removed from the current PR

psychocoderHPC reviewed (changes requested):
one open point we need to discuss.

psychocoderHPC commented on test/unit/acc/src/AccTagTest.cpp:
```diff
@@ -248,3 +252,34 @@ TEMPLATE_LIST_TEST_CASE("kernel specialization with tags", "[acc][tag]", TestAcc
+    }
+}
+
+#include <iostream>
```
please remove

SimeonEhrig committed:
implement AccIsEnabled

- the value of AccIsEnabled indicates whether an accelerator is enabled for a specific tag
- AccTags list which contains all tags
- EnabledAccTags contains all tags, where the related Acc is enabled
- isCpuTag returns true, if the related Acc is bounded the cpuplatform
- enable relaxed template template argument matching for Clang base compiler
